### DISCUSSION TITLE
Custom variables

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#146] [Ideas]
This idea actually emerged in #127, which discusses about storing shell command output values into custom variables. As custom variables are quite a big concept, and relate to other features too, I created an own discussion for it.

- A bullet point item means: Planned
- [ ] An empty box means: Will be implemented next.
- [x] A checked box means: Implemented, but not yet released.
- ✔️ means: Implemented and released.

## Features that could set values to custom variables
- #127: Will offer a new output channel that stores values into variables.
- ✔️ #17: Will store values submitted by a user to custom variables.
- A UI element (read more below)

## Features that could read the values
- ✔️ Shell command execution
- ✔️ A UI element (read more below)
- ✔️ #17: When input elements are populated, their default values can be read from variables (this can be discussed in #17).

## Naming conventions for custom variables
- ✔️ A custom variable's name begins with `{{_` and ends with `}}`. The underscore `_` prefix avoids possible name collisions between users' custom variables and all the new built-in variables that will be developed for SC in the future, because built-in variables will never use an underscore `_` as a prefix.
- ✔️ User can decide the name of the variable using letters `a-z`, numbers `0-9` and an underscore `_`.
- ✔️ Minimum lenght for the user-decideable part of the name is one character. No maximum.
- ✔️ Upper case / lower case letters do not matter, they are interpreted the same.
- Example: `{{_my_variable}}` (same as `{{_MY_VARIABLE}}` or `{{_My_VaRiAbLe}}`)
- Example: `{{_x}}` (minimun length variable name)

## User interface
- ✔️ I'm thinking about creating a right side pane element for showing a list of currently defined custom variables.
- Variable values could be altered manually in the side pane (later).

## Other features
- ✔️ Custom variables are always accessible to all shell commands, so multiple shell commands can share values by writing/reading the same variables. In programming terms, the variables could be described to be _global_.
- ✔️ Default values: if a custom variable does not yet have a value, a default value for it can be defined in each shell command's settings. Default values are shell command specific, so two shell commands can define different default values for the same variable.
- ✔️ Added 2022-03-01: In addition to default values, add an option to silently cancel execution if the variable does not have a value. Maybe not so useful for custom variables, but will be good for built-in `{{file_*}}` variables when executing shell commands via events in a situation where no file is active, in which case the execution can be automatically cancelled without error messages.
- ✔️ Each custom variable need to be defined before they can be used, so that SC will know all the available custom variables in each of the features that write or read their values. Defining a variable is not the same as setting a default value for it.
- <a name="store-on-disk">Later: Custom variables can be stored on disk, so they will remember their values when Obsidian quits. This would be an optional setting, but I haven't yet decided, will it be a global setting for all variables, or toggleable for each variable separately. The file where the variables would be stored, would be outside of the `data.json` settings, file, but I'm not yet sure about the format and location (in vault root or in plugin folder? Freely decidable name/path or not?).

## Variable types
- ✔️ Text: Allows all types of characters. **This will be implemented first**, other types later.
- Integer: Allows only numbers `0-9`, all other characters will be silently stripped away. Edit 2022-11-24: Should negative values be supported by allowing `-` prefix? Or should there be two integer types: one which supports both positive and negative values, and one that supports only positive?
- Decimal: Allows only numbers `0-9` and a decimal point (a dot `.` or comma `,`, variable will have a setting for defining if a comma will be converted into a dot, or vice-versa). There will be no support for a thousands separator. In some countries, comma is used as a thousands separator, but it won't work here.
- For both integer and decimal: Need to consider annoying values like `2.3e-5`. Should they be supported or not?
- List: Contains multiple values. Value type needs to be defined, i.e. one of the scalar types above.

**Comment by FelipeRearden** [comment]
This is gonna be amazing @Taitava!!!
